packages/safe-tools-client/app/utils/chain-connection-manager.ts

```diff
@@ -232,6 +243,12 @@ export class ChainConnectionManager {
 }
 
 class ConcreteStrategyFactory implements ConnectionManagerStrategyFactory {
+  mockProvider?: TestWalletConnectProvider;
+
+  setMockProvider(mockProvider: TestWalletConnectProvider) {
+    this.mockProvider = mockProvider;
+  }
+
```

Comment:
I find it a bit odd to have "test" code inside the production code, is it not possible to mock the walletConnect and/or metamask providers ? Or maybe we could expose the web3 on wallet service so we could set a fake provider on each tests ?

Reply:
Or if a package like this one would help https://github.com/rsksmart/mock-web3-provider

Reply:
So, the `mockProvider` in this spinnet is already a package very similar to the the one you linked - that one is cypress specific but the one I'm using is more general, it's called [eth-testing](https://github.com/VGLoic/eth-testing)

The issue here is injecting that mock provider into the right place so that it's used.

Background: This file is quite convoluted already, it makes use of a strategy factory pattern so it's quite hard to actually inject it into the right place, I think this code has been inherited and copypasted and could be cleaned up a bit.

The approach I have landed on is a medium between the first really hacky solution (put the mockProvider on the `window` object), and a really complex one I made that had a whole new strategy factory and ended up adding a lot of bloated code.

Then I noticed that this file is already switching based on the ember environment to inject a mock: https://github.com/cardstack/cardstack/blob/8b09163a0fbcceb56f277af216acc0929defdd73/packages/safe-tools-client/app/utils/chain-connection-manager.ts#L96-L99

So I figured the least confusing thing would be to follow the conventions for the file so at least there weren't two different non-ideal ways of doing it in the same file :)

I can definitely make this a lot better by restructuring the connection manager to simplify it and make it easier to inject mocked providers into it, along with just being easier to follow in general - I'm happy to do that and I think it would be a better result, the only question is if that is a yak that I should shave right now or not.

Reply:
Btw I think that probably the more idiomatic ember solution here, for both places where the mocks are used, are to use services - a localStorage service and an ethereumProvider service, and then in the tests inject mock versions of those using the ember testing tooling. That is what I would do if writing from scratch I think.